Better Sweep Metrics

gmaretic opened the pull request:
**Goals (and why)**:
Add metrics for time spent sweeping, and tagged metrics for number of cells and values deleted on table level.

**Implementation Description (bullets)**:
Added time information to sweep results and sweep progress to be able to track how much time sweep has taken across runs. Added tagged metrics for table level sweep (but haven't actually added tags yet due to unclear status).

**Concerns (what feedback would you like?)**:
This is still unfinished, so don't pay too much attention on style and the fact that tests are broken. I'm interested in whether the overall approach makes sense and any high level comments.

**Where should we start reviewing?**:
SweepMetrics.java ?

**Priority (whenever / two weeks / yesterday)**:
Today if possible

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/2672)
<!-- Reviewable:end -->

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SpecificTableSweeper.java:
```diff
@@ -201,12 +198,14 @@ private static SweepResults getCumulativeSweepResults(TableToSweep tableToSweep,
         long minimumSweptTimestamp = Math.min(
                 tableToSweep.getPreviousMinimumSweptTimestamp().orElse(Long.MAX_VALUE),
                 currentIteration.getSweptTimestamp());
+        long timeInMillis = tableToSweep.getTimeInMillisPreviously() + currentIteration.getTimeInMillis();
```
I don't think this is going to work right now - you also need to write out the cumulative time in millis to the `newProgress` value in `saveIntermediateSweepResults`.

Also need to check the `SweepProgress` can be read fine when this field isn't in the DB (when we first roll this out.

gmaretic replied:
We should probably use a new sweep progress table, to avoid breaking backwards compatibility if we go ahead with this.

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,38 +15,39 @@
+import com.palantir.tritium.metrics.registry.MetricName;
 
 @SuppressWarnings("checkstyle:FinalClass")
 public class SweepMetrics {
```
This class is doing things - counting continual sweep metrics and end-of-table sweep metrics.  Consider making 2 classes to capture these different requirements?

tboam replied:
In fact - you might even get away with a single class of which there are 2 instances - one for continual metrics and one for end-of-table.  You'd have to avoid metric name collisions and possibly some other problems but remove a lot of duplication.

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,38 +15,39 @@
+    private final TableSpecificHistogramMetric cellsSweptHistogramForTable =
+            new TableSpecificHistogramMetric("cellTimestampPairsExaminedForTable");
+    private final TableSpecificHistogramMetric cellsDeletedHistogramForTable =
+            new TableSpecificHistogramMetric("staleValuesDeleted");
```
Missing "ForTable" means this has the same name as Histogram above

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -60,16 +61,85 @@ void update(long value) {
+        return ImmutableMap.of();
+    }
+
+    void examinedCellsOneIteration(long numExamined) {
```
Why don't we tag the continuous metrics with the table name as well?

gmaretic replied:
We could, but not sure if it is valuable. This is very easy to add in the new impl.

gmaretic commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -60,16 +61,85 @@ void update(long value) {
     }
 
+    void deletedCellsFullTable(long numDeleted, TableReference tableRef) {
+        cellsDeletedForTable.update(numDeleted, tableRef);
```
Agreed, now simplified

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SpecificTableSweeper.java:
```diff
@@ -279,9 +292,17 @@ private void saveFinalSweepResults(TableToSweep tableToSweep, SweepResults final
+    private void reportSweepMetricsAfterOneIteration(SweepResults sweepResults) {
+        sweepMetrics.examinedCellsOneIteration(sweepResults.getCellTsPairsExamined());
+        sweepMetrics.deletedCellsOneIteration(sweepResults.getStaleValuesDeleted());
+        sweepMetrics.sweepTimeOneIteration(sweepResults.getTimeInMillis());
```
We should record time since start metric on each iteration too.  I know this goes up by 1 second per second so doesn't look that useful - but it allows us to create a monitor for "we've been sweeping the same table for 3 days and still haven't completed it".  If we just have this at the end of the sweep we may never see the metric if it is growing faster than we can sweep it.

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SpecificTableSweeper.java:
```diff
@@ -201,12 +199,19 @@ private static SweepResults getCumulativeSweepResults(TableToSweep tableToSweep,
+                .map(start -> new SystemClock().getTimeMillis() - start)
+                .orElse(currentIteration.getTimeInMillis());
 
         return SweepResults.builder()
```
I don't like how we're using SweepResults - it seems to be partially instantiated in a number of places.  The only way you know what is safe to call on a particular instance is to read back through the code to find out what has been set on the instance you've got which seems dodgy to me.

We should either make sure we copy all values across from one instantiation to another (i.e. when making the cumulative metrics), abandon Immutables here and force devs to use a constructor, or have separate classes for the different types of SweepResults.

gmaretic commented:
This is almost ready from my side. There are two things to consider before merging:

1. Are we still blocked from merging tagged metrics? Is there a reasonable way forward?
2. How do we deal with unsafe tables? Current naive solution treats them as one table, but that's hardly ideal. The way I see it we have two options:
a) Simple solution - use a (one way) hash function on the table name. This would solve our issue, but is inherently still unsafe as it opens us up to a brute force/dictionary attack.
b) Assign each unsafe table a random UUID. This can be cached in memory and persisted in an Atlas table. Then each time we encounter a new unsafe table, we check in the DB and if it isn't there we assign it a new UUID. Not sure if this is necessary, but probably worth it.

tboam commented:
For unsafe tables I think there's too much overhead in assigning UUIDs to tables and having to have another internal table to track these, remove UUIDs when tables are deleted etc.

One way hash is probably fine but we'd have to understand what's an appropriate level of encryption.

Lumping them all into 1 table is probably going to be the approach for now and then we can follow up.

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,62 +15,99 @@
-        MeterMetric(String name) {
-            this.meter = metricsManager.registerOrGetMeter(SweepMetrics.class, null, name);
+        public void setValue(long newValue) {
+            value = newValue;
```
This isn't thread safe because writing the long might not be an atomic operation.  You should make this either volatile or an AtomicLong.

gmaretic replied:
Well spotted!

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,62 +15,99 @@
 
-        private String aggregateMetric() {
-            return MetricRegistry.name(SweepMetrics.class, name);
+        void updateAggregate(long value) {
```
This is just used for the time elapsed so I would rename to make this clear

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,62 +15,99 @@
-    private final MeterMetric cellsSweptMeter = new MeterMetric("cellsSwept");
-    private final MeterMetric cellsDeletedMeter = new MeterMetric("cellsDeleted");
-    private final MeterMetric sweepErrorMeter = new MeterMetric("sweepError");
+    private final SweepMetric cellsSweptMetric = new SweepMetric("cellTimestampPairsExamined");
```
I'm not entirely sure how tagged metrics work but I think there's a problem with this approach.  Both the incremental and end of table metrics will have the same name.  We can filter by tag to see the end of table results, but we can't filter out the end of table results from the then incremental (untagged) results.

When we create sweep metrics we should pass in whether it is incremental or full results and the table name and register the appropriate metrics.  This may also let us tag the incremental results with a table name.

We can ask the MetricsRegistry for the Metric object each time if we don't want to store a map of them.

gmaretic replied:
Alternatively we can add more tags as I've done now. I feel that's way cleaner, but feel free to push back.

gmaretic commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/SweepResults.java:
```diff
@@ -55,8 +67,10 @@ public SweepResults accumulateWith(SweepResults other) {
+                .sweptTimestamp(Math.min(getSweptTimestamp(), other.getSweptTimestamp()))
                 .nextStartRow(other.getNextStartRow())
+                .timeInMillis(getTimeInMillis() + other.getTimeInMillis())
+                .timeSweepStarted(getTimeSweepStarted())
```
Done

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SpecificTableSweeper.java:
```diff
@@ -195,19 +192,7 @@ private void processSweepResults(TableToSweep tableToSweep, SweepResults current
-                .sweptTimestamp(minimumSweptTimestamp)
-                .nextStartRow(currentIteration.getNextStartRow())
-                .build();
+        return tableToSweep.getPreviousSweepResults().accumulateWith(currentIteration);
```
Love it when a big chunk of code disappears like this - good indicator we've done it right.

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepMetrics.java:
```diff
@@ -15,62 +15,106 @@
-    void examinedCells(long numExamined) {
-        cellsSweptHistogram.update(numExamined);
-        cellsSweptMeter.update(numExamined);
+    void updateMetricsOneIteration(SweepResults sweepResults) {
```
Few thoughts on this class still, it isn't sitting right with me.  I think we should tag the `OneIteration` metrics with the table name as well - we have the tableRef at the call site and it will allow us to compare table1 vs table 2 or table1 at customer1 vs table1 at customer 2.

I'm also not sure it is possible to plot a metric where a tag isn't present in DD.  I can filter to a single tag value, but I can't say "give me these metric values where this tag isn't specified".

Secondly, I still think we should have separate metrics for incremental and end-of-table metrics.  These are updated at different rates, I wouldn't want to plot incremental and end-of-table as a single metric or compare them against each other.  They are designed to tell us different things and create different monitors.


There's a also 'code smell' in this class where we have the same code path duplicated.

gmaretic commented:
@tboam I have discussed what metrics made sense with @jeremyk-91 and reworked what we're tracking:

- for cells examined, deleted and time spent sweeping: a non-tagged (by table name, they are all tagged) meter each iteration to show current sweep metrics, a tagged histogram each iteration to have per-table statistics on these values, and a non-tagged histogram after each table is fully swept to have statistics across tables.

- for time elapsed: a non-tagged currentValue metric each iteration to have data on how much time has elapsed since we started sweeping the current table, and a non-tagged histogram after each table is fully swept to have statistics on how much sweep took across tables.

- for sweep errors - just a non-tagged meter.

Other than these, it didn't seem valuable to include others: we have too sparse data for tagged metrics tracked only after a table is fully swept, and there is no point in having metrics that are just the aggregates of the tagged metrics since we should be able to aggregate ourselves.

I also refactored the class into a new package and hopefully made it easier to follow. If we decide to add new Meters or change the metrics, it should be super-easy (which we can consider once we see how useful the current metrics are)

fsamuel-bs commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/SweepMetricsFactory.java:
```diff
@@ -0,0 +1,113 @@
+        return new SweepMetricsFactory.ListOfMetrics(
+                createMeter(name, UpdateEvent.ONE_ITERATION, false),
+                createHistogram(name, UpdateEvent.ONE_ITERATION, true),
+                createHistogram(name, UpdateEvent.FULL_TABLE, false));
```
I think we use ExponentiallyDecayingReservoir for histograms. It's docs follow:

```
    /**
     * Creates a new {@link ExponentiallyDecayingReservoir} of 1028 elements, which offers a 99.9%
     * confidence level with a 5% margin of error assuming a normal distribution, and an alpha
     * factor of 0.015, which heavily biases the reservoir to the past 5 minutes of measurements.
     */
```

1. We saw iterations of sweep having an interval of 7s between each other. Do we want histograms for each iteration?
2. Since sweeping two full tables can take a while, do we also want histograms for `UpdateEvent.FULL_TABLE` kinds of event?

fsamuel-bs commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/TagAndLabel.java:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+
```
Random empty lines appear.

fsamuel-bs commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/SweepTaggedMetric.java:
```diff
@@ -0,0 +1,63 @@
+            TableReference tableRef, Boolean taggedWithTableName) {
+        ImmutableMap.Builder<String, String> tagsBuilder = ImmutableMap.builder();
+        tagsBuilder.put(metricType.getTag(), metricType.getLabel());
+        tagsBuilder.put(updateEvent.getTag(), updateEvent.getLabel());
```
I think I'd rather have the event type as part of the name of the metric. As in, I don't think the aggregate across event types makes sense per se, so if we're using just one of the metrics at a time, we'd rather explicitly make them separate metrics.

fsamuel-bs commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/SweepTaggedMetric.java:
```diff
@@ -0,0 +1,63 @@
+    private static Map<String, String> constructTags(MetricType metricType, UpdateEvent updateEvent,
+            TableReference tableRef, Boolean taggedWithTableName) {
+        ImmutableMap.Builder<String, String> tagsBuilder = ImmutableMap.builder();
+        tagsBuilder.put(metricType.getTag(), metricType.getLabel());
```
I don't know if we need to put the metricType in the tags. I think histograms will generate p50, p95 and p99, and markers will generate counts. I may be wrong here though. It'd be a problem if histograms also generated counts, but I don't think we want the histograms anyway...

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/SweepMetricsFactory.java:
```diff
@@ -0,0 +1,115 @@
+        private final List<SweepMetric> metricsList;
+
+        ListOfMetrics(SweepMetric... metrics) {
+            this.metricsList = Arrays.stream(metrics).collect(Collectors.toList());
```
`Arrays.asList(metrics)` is simpler

tboam commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/UpdateEvent.java:
```diff
@@ -0,0 +1,38 @@
+    public abstract String getTag();
+    public abstract String getLabel();
+
+    public static final UpdateEvent ONE_ITERATION = ImmutableUpdateEvent.builder()
```
I still disagree with using tags for iteration vs full table metrics.  By default this is going to average both values on each graph.  I expect the iteration metrics to be insignificant compared to the full table metrics.  I can't think of a use case for combining the two metrics types at render time, or wanting to plot them against each other.  I think this is going to catch us and other users out.

Having them as separate metrics avoids all of this confusion.

gmaretic replied:
That makes perfect sense, I misunderstood how tagged metrics worked. Going to change this.

gmaretic commented on atlasdb-client/src/main/java/com/palantir/atlasdb/util/MeanValueMetric.java:
```diff
@@ -0,0 +1,45 @@
+            mean = (double) entry;
+        }
+        else {
+            double normalizedEntry = entry / numberOfEntries;
```
I am fine keeping a running tally if that is clearer, but this is definitely correct:
avg(x_1..x_{n+1}) = 
= (x_1 + ... + x_n + x_{n+1}) / (n + 1) =
= ((x_1 + ... + x_n + x_{n+1}) / n) * n / (n + 1) =
= ((x_ 1 + ... + x_n) / n + x_{n+1} / n) * n / (n + 1) =
= (avg(x_1...x_n) + x_{n+1} / n) / (n + 1) * n 
(swapped order of division and multiplication for more precision in floating point arithmetic)

Can add tests or rewrite

gmaretic commented on atlasdb-client/src/main/java/com/palantir/atlasdb/util/CurrentValueMetric.java:
```diff
@@ -0,0 +1,41 @@
+        value = newValue;
+    }
+
+    public static class MaximumValueMetric extends CurrentValueMetric {
```
Makes sense, trying to be too fancy

gmaretic commented:
Now we register non-tagged metrics for the metrics that are not tagged with the table name. This allows us to register histograms with HdrHistogramReservoir, so avoids custom metrics for full table sweeps.
I plan to improve MetricsManager in a follow-up PR and then refactor this to use MetricsManager instead of MetricRegistry and TaggedMetricRegistry directly.

jeremyk-91 commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/SweepMetricsManager.java:
```diff
@@ -0,0 +1,50 @@
+import com.palantir.atlasdb.keyvalue.api.SweepResults;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+
+@SuppressWarnings("checkstyle:FinalClass")
```
nit: I'm guessing this is for Mockito? Would be good to leave a comment

gmaretic replied:
It was, but since this was not mocked anymore, might as well make it final. If someone desperately wants to extend, we can change it later :)

jeremyk-91 commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/metrics/UpdateEvent.java:
```diff
@@ -0,0 +1,33 @@
+
+package com.palantir.atlasdb.sweep.metrics;
+
+public enum UpdateEvent {
```
nit: This feels more like an `UpdateEventType` as opposed to an `UpdateEvent`, though feel free to ignore this

gmaretic commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/SweepResults.java:
```diff
@@ -34,46 +36,76 @@
+            return fst;
+        } else {
+            return snd;
+        }
```
I think this is awesomer, since the (now fixed) method name already says what it does

gmaretic commented on atlasdb-client/src/main/java/com/palantir/atlasdb/util/CurrentValueMetric.java:
```diff
@@ -0,0 +1,32 @@
+
+    public void setValue(long newValue) {
+        value = newValue;
+    }
```
Sure, I was going for "if someone needs it, they can parameterise", but seeing as it's so simple, might as well

tboam commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -50,8 +50,25 @@ develop
+         - Sweep metrics were reworked. Sweep now exposes metrics indicating the total number of cells examined, cells deleted, time spent sweeping, and time elapsed since sweep started on the current table that are updated after each iteration of sweep and separate metrics that are updated after each table is fully swept.
+           Additionally, sweep now exposes metrics tagged with table names that expose the total number of cells examined, cells deleted, time spent sweeping per iteration for each table separately.
+           Logs will also include the new timing information.
+           Sweep now exposes the following metrics with the common prefix ``com.palantir.atlasdb.sweep.metrics.SweepMetric.``:
```
Which metrics (if any) are no longer reported?  Once this is merged and running somewhere you should upgrade sweep metrics dashboard to report either the old metric or the relevant new one.

tboam commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -50,8 +50,25 @@ develop
+           Sweep now exposes the following metrics with the common prefix ``com.palantir.atlasdb.sweep.metrics.SweepMetric.``:
+
+              - ``cellTimestampPairsExamined.meter.perIteration``
+              - ``cellTimestampPairsExamined.histogram.perTable``
```
I expect a `perTable` metric to be broken down (tagged) by table.  If it is easy can we rename to `currentTable` and `currentIteration`?

tboam commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -50,8 +50,25 @@ develop
+              - ``sweepTimeSweeping.histogram.perTable``
+              - ``sweepTimeSweeping.histogram.perIteration`` (tagged)
+              - ``sweepTimeElapsedSinceStart.currentValue.perIteration``
+              - ``sweepTimeElapsedSinceStart.histogram.perTable``
```
My intuition is that any of the `histogram.perTable` metrics are going to be hard to reason about.  Happy to include for now and then look at the graphs, but if they aren't answering a specific question then we should remove them.

tboam commented:
Ship it! 🚢
